### DISCUSSION TITLE
columnSearch return response after second search

### DISCUSSION THREAD

**Discussion by vipinrao2022** [rappasoft/laravel-livewire-tables#1348] [Q&A]
Hello, I have been struggling with columnSearch whenever I type the char on columnSearch, not a global search. The rappasoft doesn't return the request instead it returns the request in a second char search. Meaning that the value is not passed in the first search.

**Comment by lrljoe** [comment]
Can you do a screen capture, and share your table component code please?

And let me know versions (package, laravel, livewire, alpinejs), and which theme you're using please

**Comment by vipinrao2022** [comment]
yea sure

**Comment by vipinrao2022** [comment]
![Screenshot from 2023-09-07 13-19-14](https://github.com/rappasoft/laravel-livewire-tables/assets/133568134/3385a7d0-f684-46e3-b35c-6b8b534742e7)
![Screenshot from 2023-09-07 13-20-18](https://github.com/rappasoft/laravel-livewire-tables/assets/133568134/a2fc4c6a-7f99-43ef-ad00-174bd1a3ac0e)
b71499e)
![Screenshot from 2023-09-07 13-22-51](https://github.com/rappasoft/laravel-livewire-tables/assets/133568134/861d5f65-c987-45e8-88e0-ec1b114b81a3)

![Screenshot from 2023-09-07 13-19-46](https://github.com/rappasoft/laravel-livewire-tables/assets/133568134/b2b6c872-de70-4183-ac03-8d4986b5d9c9)

Laravel version is 8
php version is 7.4
rapasoft version is 2.14
alpine is 3.1
livewire 2.12
I have attached the pics of the code. Kindly do help :)

**Reply by lrljoe** [reply]
Okay!

It's an oddity, but when you start putting filters into the builder() function, rather than the dedicated filter function, it does (sporadically) end up a lifecycle behind where it should be.

You'll have far better success if you use the filters() and a filter() callback within there.

I'm more active on Discord at the moment, but if you share a few examples of what's in your if statements then I can probably give you some better pointers!

**Reply by vipinrao2022** [reply]
Yeah .. I know the filter will work. But I want to try with search, & thing abt if statements, I have removed all the ifs and kept only the below statement inside builder but also the first request doesn't come to the builder only the second request comes and returns the result.

```
public function builder(): Builder
    {

            return TblInsuredPartyModel::query()
            ->when($this->columnSearch['PracticeDataSource'] ?? null, fn ($query, $PracticeDataSource) => $query->where('tbl_insured_parties.PracticeDataSource', 'like', '%' . $PracticeDataSource . '%'))
            ->when($this->columnSearch['id'] ?? null, fn ($query, $id) => $query->where('tbl_insured_parties.id', 'like', '%' . $id . '%'));

     }
```

**Reply by lrljoe** [reply]
The 'when' statements at present fire unexpectedly on v2, often between filter/search values updating.  You may have some luck with using a setSearchDebounce approach.

I'm pushing some v3 improvements through, but it doesn't fully address all scenarios of this.

The issue is that the filter/search values aren't available as early in the lifecycle as they should be in v2 (Livewire v2), so when your builder() is called, it isn't as clean and aware as it should be.

You absolutely can use the searchable() callbacks on the columns in the meantime.

I'll be backporting fixes into v2, and tidying over the next week or two, and publishing roadmaps for the different branches.

In the meantime, you can definitely call the functions prematurely to avoid the lifecycle issue, so long as you are using a trait, as you can then hook in to the relevant events.

I am actively reviewing events in v3 with a view to introducing native hooks, as well as smoother guidance for the various features.

If you poke me on Discord then you're likely to get an interim solution fairly quickly, as I am more active on there at the mo

**Reply by vipinrao2022** [reply]
Okay sure, thank you. I'll Look at that..how can I connect you on Discord...? I mean link or username..?

**Reply by lrljoe** [reply]
Link should be pinned to the discussions on github!

Let me know if it has disappeared 

https://discord.gg/mkmEfq7hUC

**Reply by vipinrao2022** [reply]
Okay.. Thank You...

**Reply by lrljoe** [reply]
Did it all make sense in the end in terms of column searches and not doing it in the builder() function?

**Reply by vipinrao2022** [reply]
I didn't get you, sorry. Can you please elaborate?

**Reply by lrljoe** [reply]
Use a callback on the column rather than when in the builder,

```
Column::make('Banana Name')
    ->searchable(
        fn(Builder $query, $searchTerm) => $query->orWhere('bananas.name', $searchTerm)
    ),
```

You can use the sortable or searchable callbacks on both standard and label cols.

Reach out on the official discord if this still isn't clear. I'm more available there at the mo
